test/select_io_dset.c

```diff
@@ -782,6 +820,665 @@ test_cmpd_with_bkg(hid_t fid, unsigned chunked)
+ *
+ *  Case b--setting for read from ndsets:
+ *    Datatype for all datasets: H5T_NATIVE_LONG
+ */
+static herr_t
+test_multi_dsets_no_bkg(hid_t fid, unsigned chunked, unsigned dtrans)
+{
+    size_t  ndsets;
+    int     i, j;
+    hid_t   dcpl        = H5I_INVALID_HID;
+    hid_t   dxpl        = H5I_INVALID_HID;
+    hid_t   ntrans_dxpl = H5I_INVALID_HID;
+    hsize_t dims[1];
+    hsize_t cdims[1];
+
+    hid_t file_sids[MULTI_NUM_DSETS];
+    hid_t mem_sids[MULTI_NUM_DSETS];
+    hid_t mem_tids[MULTI_NUM_DSETS];
+
+    char  dset_names[MULTI_NUM_DSETS][DSET_NAME_LEN];
+    hid_t dset_dids[MULTI_NUM_DSETS];
+
+    size_t buf_size;
+    int   *total_wbuf       = NULL;
+    int   *total_trans_wbuf = NULL;
+    int   *total_rbuf       = NULL;
+    long  *total_lrbuf      = NULL;
+
+    int  *wbufi[MULTI_NUM_DSETS];
+    int  *trans_wbufi[MULTI_NUM_DSETS];
+    int  *rbufi[MULTI_NUM_DSETS];
+    long *l_rbufi[MULTI_NUM_DSETS];
+
+    const void *wbufs[MULTI_NUM_DSETS];
+    void       *rbufs[MULTI_NUM_DSETS];
+    const char *expr = "2*x";
+
+    ndsets = MAX(MULTI_MIN_DSETS, MULTI_NUM_DSETS);
+
+    dims[0] = DSET_SELECT_DIM;
+
+    if ((dcpl = H5Pcreate(H5P_DATASET_CREATE)) < 0)
+        FAIL_STACK_ERROR;
+
+    if (chunked) {
+        cdims[0] = DSET_SELECT_CHUNK_DIM;
+        if (H5Pset_chunk(dcpl, 1, cdims) < 0)
+            FAIL_STACK_ERROR;
+    }
+
+    /* Create dataset transfer property list */
+    if ((dxpl = H5Pcreate(H5P_DATASET_XFER)) < 0)
+        FAIL_STACK_ERROR;
+
+    if ((ntrans_dxpl = H5Pcopy(dxpl)) < 0)
+        FAIL_STACK_ERROR;
+
+    /* Set data transform */
+    if (dtrans)
+        if (H5Pset_data_transform(dxpl, expr) < 0)
+            FAIL_STACK_ERROR;
+
+    /* Set up file space ids, mem space ids, and dataset ids */
+    for (i = 0; i < (int)ndsets; i++) {
+        if ((file_sids[i] = H5Screate_simple(1, dims, NULL)) < 0)
+            FAIL_STACK_ERROR;
+
+        if ((mem_sids[i] = H5Screate_simple(1, dims, NULL)) < 0)
+            FAIL_STACK_ERROR;
+
+        /* Create ith dataset */
+        if (chunked) {
+            if (snprintf(dset_names[i], DSET_NAME_LEN,
+                         dtrans ? "multi_chk_trans_dset%u" : "multi_chk_ntrans_dset%u", i) < 0)
+                FAIL_STACK_ERROR;
+        }
+        else {
+            if (snprintf(dset_names[i], DSET_NAME_LEN,
+                         dtrans ? "multi_contig_trans_dset%u" : "multi_contig_ntrans_dset%u", i) < 0)
+                FAIL_STACK_ERROR;
+        }
+
+        /* dset0: H5T_NATIVE_INT, the other datasets: H5T_NATIVE_LLONG or H5T_NATIVE_SHORT */
+        if (i == 0) {
+            if ((dset_dids[i] = H5Dcreate2(fid, dset_names[i], H5T_NATIVE_INT, file_sids[i], H5P_DEFAULT,
+                                           dcpl, H5P_DEFAULT)) < 0)
+                FAIL_STACK_ERROR;
+        }
+        else {
+            if ((dset_dids[i] =
+                     H5Dcreate2(fid, dset_names[i], ((HDrandom() % 2) ? H5T_NATIVE_LLONG : H5T_NATIVE_SHORT),
```

Comment:
Maybe use int16_t, int32_t, and int64_t to ensure ype conversions and size changes